rankwatch-projects/src/main.js

```diff
@@ -0,0 +1,303 @@
+        label: 'Ranked URL',
+        dataType: 'STRING',
+        semantics: {
+          conceptType: 'DIMENSION'
+        }
+      },
+      {
+        name: 'url',
+        label: 'URL',
+        dataType: 'STRING',
+        semantics: {
+          conceptType: 'DIMENSION'
+        }
+      },
+      {
+        name: 'p_id',
+        label: 'Project ID',
+        dataType: 'STRING',
+        semantics: {
+          conceptType: 'DIMENSION'
+        }
+      },
+      {
+        name: 'initial_rank',
+        label: 'Initial Rank',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'current_rank',
+        label: 'Current Rank',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'highest_rank',
+        label: 'Highest Rank',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'search_engine_id',
+        label: 'Search Engine ID',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'DIMENSION'
+        }
+      },
+      {
+        name: 'page_score',
+        label: 'Page Score',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'search_volume',
+        label: 'Search Volume',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'page_number',
+        label: 'Page Number',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'previous_rank',
+        label: 'Previous Rank',
+        dataType: 'NUMBER',
+        semantics: {
+          conceptType: 'METRIC'
+        }
+      },
+      {
+        name: 'added_on',
+        label: 'Added On',
+        dataType: 'STRING',
+        semantics: {
+          conceptType: 'DIMENSION',
+          semanticGroup: 'DATE_OR_TIME',
+          semanticType: 'YEAR_MONTH_DAY'
+        }
+      }
+    ]
+  };
+  return s;
+}
+
+// https://developers.google.com/datastudio/connector/reference#getdata
+function getData(request) {
+  var cc = DataStudioApp.createCommunityConnector();
+  var config = cc.getConfig();
+
+  var dataSchema = [];
+  var fixedSchema = getSchema().schema;
+  request.fields.forEach(function(field) {
+    for (var i = 0; i < fixedSchema.length; i++) {
+      if (fixedSchema[i].name == field.name) {
+        dataSchema.push(fixedSchema[i]);
+        break;
+      }
+    }
+  });
+
+  var data = [];
+
+  for (var j = 0; j < 10; j++) {
+    var offset = parseInt(500 * j);
+
+    var userProperties = PropertiesService.getUserProperties();
+    username = userProperties.getProperty('dscc.username');
+    password = userProperties.getProperty('dscc.password');
+
+    var rawResponse = UrlFetchApp.fetch(
```

Comment:
Since you make a lot of calls to the rankwatch api w/ the authorization headers, it's probably worth refactoring out this logic into one function.

Reply:
I have moved the URLFetchApp method to other function, moving the entire code inside the loop will adversely affect by consuming more CPU cycles as the referenced variable ```data``` will be accessed repeatedly and the call event will happen unnecessarily. Most of the users may do not have so much data that the loop would more than once! This is just a precaution for future, in case a large account uses this. If you insist I can do it.